src/rgw/rgw_aws_lambda.cc

```diff
@@ -0,0 +1,116 @@
+
+namespace lambda {
+    const char *ALLOCATION_TAG = "TEST";
+    static std::shared_ptr<Aws::Lambda::LambdaClient> client;
+
+}
+
+#include <boost/lockfree/queue.hpp>
+#include <string>
+#include <utility>
+#include <aws/sns/SNSClient.h>
+#include <aws/sns/model/PublishRequest.h>
+
+namespace lambda {
+
+    static const int STATUS_OK = 1;
+    static const int STATUS_FULL = 2;
+    static const int STATUS_MANAGER_STOPPED = 3;
+
+    struct message_t {
+        std::string function_name;
+        std::string payload;
+
+        message_t(std::string function_name, std::string payload) : function_name(std::move(function_name)),
+                                                                    payload(std::move(payload)) {};
+    };
+
+    typedef boost::lockfree::queue<message_t *, boost::lockfree::fixed_sized<true>> MessageQueue;
+
+    class Manager {
+    private:
+        MessageQueue messageQueue;
+        std::thread runner;
+        bool stopped;
+
+        void publish_internal(message_t *msg) {
+          std::cout << "Invoking.." << std::endl;
+          Aws::Client::ClientConfiguration configuration;
+          configuration.verifySSL = true;
+          configuration.connectTimeoutMs = 30000;
+          client = Aws::MakeShared<Aws::Lambda::LambdaClient>(ALLOCATION_TAG, configuration);
+
+          Aws::Lambda::Model::InvokeRequest invokeRequest;
+          invokeRequest.SetInvocationType(Aws::Lambda::Model::InvocationType::RequestResponse);
+          invokeRequest.SetLogType(Aws::Lambda::Model::LogType::Tail);
+          invokeRequest.SetFunctionName(Aws::String(msg->function_name));
+          std::shared_ptr<Aws::IOStream> body = Aws::MakeShared<Aws::StringStream>(ALLOCATION_TAG);
+          *body << msg->payload;
+          invokeRequest.SetBody(body);
+          invokeRequest.SetContentType("application/json");
+          auto response = client->Invoke(invokeRequest);
+          if (!response.IsSuccess()) {
+            std::cout << response.GetError().GetMessage() << std::endl;
+          } else {
+            Aws::Lambda::Model::InvokeResult &result = response.GetResult();
+            Aws::IOStream &pl = result.GetPayload();
+            Aws::String returnedData;
+            std::getline(pl, returnedData);
+            std::cout << "Success " << returnedData << "\n" << std::endl;
+          }
+        }
+
+//    Works on messages in the queque
+        void run() {
+          while (!stopped) {
+            const auto count = messageQueue.consume_all(
+                    std::bind(&Manager::publish_internal, this, std::placeholders::_1));
+          }
+        }
+
+    public:
+        Manager(size_t max_queue_size) : messageQueue(max_queue_size), runner(&Manager::run, this), stopped(false) {}
+
+        int publish(const std::string &payload, const std::string &function_name) {
+          if (stopped)
+            return STATUS_MANAGER_STOPPED;
+          if (messageQueue.push(new message_t(function_name, payload)))
```

Comment:
who owns the message pointer?
looks like a leak. probably should set it into a unique_ptr when popped out of the queue in the `publish_internal` function